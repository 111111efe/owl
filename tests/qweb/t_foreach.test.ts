import { renderToString, snapshotTemplateCode, TestTemplateSet } from "../helpers";

// -----------------------------------------------------------------------------
// t-foreach
// -----------------------------------------------------------------------------

describe("t-foreach", () => {
  test("simple iteration", () => {
    const template = `<t t-foreach="[3, 2, 1]" t-as="item"><t t-esc="item"/></t>`;
    snapshotTemplateCode(template);
    expect(renderToString(template)).toBe("321");
  });

  test("simple iteration (in a node)", () => {
    const template = `
        <div>
          <t t-foreach="[3, 2, 1]" t-as="item"><t t-esc="item"/></t>
        </div>`;
    snapshotTemplateCode(template);
    expect(renderToString(template)).toBe("<div>321</div>");
  });

  test("iterate on items", () => {
    const template = `
        <div>
          <t t-foreach="[3, 2, 1]" t-as="item">
            [<t t-esc="item_index"/>: <t t-esc="item"/> <t t-esc="item_value"/>]
          </t>
        </div>`;
    snapshotTemplateCode(template);
    expect(renderToString(template)).toBe("<div> [0: 3 3]  [1: 2 2]  [2: 1 1] </div>");
  });

  test("iterate on items (on a element node)", () => {
    const template = `
        <div>
          <span t-foreach="[1, 2]" t-as="item" t-key="item"><t t-esc="item"/></span>
        </div>`;
    snapshotTemplateCode(template);
    const expected = `<div><span>1</span><span>2</span></div>`;
    expect(renderToString(template)).toBe(expected);
  });

  test("iterate, position", () => {
    const template = `
        <div>
          <t t-foreach="Array(5)" t-as="elem">
            -<t t-if="elem_first"> first</t><t t-if="elem_last"> last</t> (<t t-esc="elem_index"/>)
          </t>
        </div>`;
    snapshotTemplateCode(template);
    const expected = `<div> - first (0)  - (1)  - (2)  - (3)  - last (4) </div>`;
    expect(renderToString(template)).toBe(expected);
  });

  test("iterate, dict param", () => {
    const template = `
        <div>
          <t t-foreach="value" t-as="item">
            [<t t-esc="item_index"/>: <t t-esc="item"/> <t t-esc="item_value"/>]
          </t>
        </div>`;
    snapshotTemplateCode(template);
    const expected = `<div> [0: a 1]  [1: b 2]  [2: c 3] </div>`;
    const context = { value: { a: 1, b: 2, c: 3 } };
    expect(renderToString(template, context)).toBe(expected);
  });

  test("does not pollute the rendering context", () => {
    const template = `
        <div>
          <t t-foreach="[1]" t-as="item"><t t-esc="item"/></t>
        </div>`;
    snapshotTemplateCode(template);
    const context = { __owl__: {} };
    renderToString(template, context);
    expect(Object.keys(context)).toEqual(["__owl__"]);
  });

  test("t-foreach in t-foreach", () => {
    const template = `
        <div>
          <t t-foreach="numbers" t-as="number">
            <t t-foreach="letters" t-as="letter">
              [<t t-esc="number"/><t t-esc="letter"/>]
            </t>
          </t>
        </div>`;

    snapshotTemplateCode(template);

    const context = { numbers: [1, 2, 3], letters: ["a", "b"] };
    const expected = "<div> [1a]  [1b]  [2a]  [2b]  [3a]  [3b] </div>";
    expect(renderToString(template, context)).toBe(expected);
  });

  test("t-call without body in t-foreach in t-foreach", () => {
    const templateSet = new TestTemplateSet();
    const sub = `
        <t>
          <t t-set="c" t-value="'x' + '_' + a + '_'+ b" />
          [<t t-esc="a" />]
          [<t t-esc="b" />]
          [<t t-esc="c" />]
        </t>`;

    const main = `
        <div>
          <t t-foreach="numbers" t-as="a">
            <t t-foreach="letters" t-as="b">
              <t t-call="sub" />
            </t>
            <span t-esc="c"/>
          </t>
          <span>[<t t-esc="a" />][<t t-esc="b" />][<t t-esc="c" />]</span>
        </div>`;

    templateSet.add("sub", sub);
    templateSet.add("main", main);
    snapshotTemplateCode(sub);
    snapshotTemplateCode(main);

    const context = { numbers: [1, 2, 3], letters: ["a", "b"] };
    const expected =
      "<div> [1] [a] [x_1_a]  [1] [b] [x_1_b] <span></span> [2] [a] [x_2_a]  [2] [b] [x_2_b] <span></span> [3] [a] [x_3_a]  [3] [b] [x_3_b] <span></span><span>[][][]</span></div>";
    expect(templateSet.renderToString("main", context)).toBe(expected);
  });

  test("t-call with body in t-foreach in t-foreach", () => {
    const templateSet = new TestTemplateSet();
    const sub = `
        <t>
          [<t t-esc="a" />]
          [<t t-esc="b" />]
          [<t t-esc="c" />]
        </t>`;

    const main = `
        <div>
          <t t-foreach="numbers" t-as="a">
            <t t-foreach="letters" t-as="b">
              <t t-call="sub" >
                <t t-set="c" t-value="'x' + '_' + a + '_'+ b" />
              </t>
            </t>
            <span t-esc="c"/>
          </t>
          <span>[<t t-esc="a" />][<t t-esc="b" />][<t t-esc="c" />]</span>
        </div>`;

    templateSet.add("sub", sub);
    templateSet.add("main", main);
    snapshotTemplateCode(sub);
    snapshotTemplateCode(main);

    const context = { numbers: [1, 2, 3], letters: ["a", "b"] };
    const expected =
      "<div> [1] [a] [x_1_a]  [1] [b] [x_1_b] <span></span> [2] [a] [x_2_a]  [2] [b] [x_2_b] <span></span> [3] [a] [x_3_a]  [3] [b] [x_3_b] <span></span><span>[][][]</span></div>";
    expect(templateSet.renderToString("main", context)).toBe(expected);
  });

  test("throws error if invalid loop expression", () => {
    const test = `<div><t t-foreach="abc" t-as="item"><span t-key="item_index"/></t></div>`;
    expect(() => renderToString(test)).toThrow("Invalid loop expression");
  });

  test.skip("warn if no key in some case", () => {
    const consoleWarn = console.warn;
    console.warn = jest.fn();

    const template = `
        <div>
          <t t-foreach="[1, 2]" t-as="item">
            <span><t t-esc="item"/></span>
          </t>
        </div>`;
    renderToString(template);
    expect(console.warn).toHaveBeenCalledTimes(1);
    expect(console.warn).toHaveBeenCalledWith(
      "Directive t-foreach should always be used with a t-key! (in template: 'test')"
    );
    console.warn = consoleWarn;
  });

  test("multiple calls to t-raw", () => {
    const templateSet = new TestTemplateSet();
    const sub = `
        <div>
          <t t-raw="0"/>
          <div>Greeter</div>
          <t t-raw="0"/>
        </div>`;

    const main = `
        <div>
          <t t-call="sub">
            <span>coucou</span>
          </t>
        </div>`;

    templateSet.add("sub", sub);
    templateSet.add("main", main);
    snapshotTemplateCode(sub);
    snapshotTemplateCode(main);
    const expected =
      "<div><div><span>coucou</span><div>Greeter</div><span>coucou</span></div></div>";
    expect(templateSet.renderToString("main")).toBe(expected);
  });
});
