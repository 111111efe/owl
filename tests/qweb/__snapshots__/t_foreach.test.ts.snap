// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`t-foreach does not pollute the rendering context 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
  }
  
  return ctx => {
    const b1 = new Block1();
    const [v2, k2, l2] = getValues([1]);
    const b3 = b1.children[0] = new BCollection(l2);
    ctx = Object.create(ctx);
    for (let i1 = 0; i1 < l2; i1++) {
      ctx[\`item\`] = v2[i1];
      ctx[\`item_first\`] = i1 === 0;
      ctx[\`item_last\`] = i1 === v2.length - 1;
      ctx[\`item_index\`] = i1;
      ctx[\`item_value\`] = k2[i1];
      b3.children[i1] = new BText(ctx['item'])
    }
    return b1;
  }
}"
`;

exports[`t-foreach iterate on items (on a element node) 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
  }
  
  class Block2 extends BNode {
    static el = elem(\`<span><owl-text/></span>\`);
    data = new Array(1);
    update() {
      this.el.firstChild.textContent = toString(this.data[0]);
    }
  }
  
  return ctx => {
    const b1 = new Block1();
    const [v2, k2, l2] = getValues([1,2]);
    const b3 = b1.children[0] = new BCollection(l2);
    ctx = Object.create(ctx);
    for (let i1 = 0; i1 < l2; i1++) {
      ctx[\`item\`] = v2[i1];
      ctx[\`item_first\`] = i1 === 0;
      ctx[\`item_last\`] = i1 === v2.length - 1;
      ctx[\`item_index\`] = i1;
      ctx[\`item_value\`] = k2[i1];
      const k4 = ctx['item'];
      const b5 = b3.children[i1] = new Block2();
      b5.data[0] = ctx['item'];
    }
    return b1;
  }
}"
`;

exports[`t-foreach iterate on items 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
  }
  
  return ctx => {
    const b1 = new Block1();
    const [v2, k2, l2] = getValues([3,2,1]);
    const b3 = b1.children[0] = new BCollection(l2);
    ctx = Object.create(ctx);
    for (let i1 = 0; i1 < l2; i1++) {
      ctx[\`item\`] = v2[i1];
      ctx[\`item_first\`] = i1 === 0;
      ctx[\`item_last\`] = i1 === v2.length - 1;
      ctx[\`item_index\`] = i1;
      ctx[\`item_value\`] = k2[i1];
      const b4 = b3.children[i1] = new BMulti(7);
      b4.children[0] = new BText(\` [\`)
      b4.children[1] = new BText(ctx['item_index'])
      b4.children[2] = new BText(\`: \`)
      b4.children[3] = new BText(ctx['item'])
      b4.children[4] = new BText(\` \`)
      b4.children[5] = new BText(ctx['item_value'])
      b4.children[6] = new BText(\`] \`)
    }
    return b1;
  }
}"
`;

exports[`t-foreach iterate, dict param 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
  }
  
  return ctx => {
    const b1 = new Block1();
    const [v2, k2, l2] = getValues(ctx['value']);
    const b3 = b1.children[0] = new BCollection(l2);
    ctx = Object.create(ctx);
    for (let i1 = 0; i1 < l2; i1++) {
      ctx[\`item\`] = v2[i1];
      ctx[\`item_first\`] = i1 === 0;
      ctx[\`item_last\`] = i1 === v2.length - 1;
      ctx[\`item_index\`] = i1;
      ctx[\`item_value\`] = k2[i1];
      const b4 = b3.children[i1] = new BMulti(7);
      b4.children[0] = new BText(\` [\`)
      b4.children[1] = new BText(ctx['item_index'])
      b4.children[2] = new BText(\`: \`)
      b4.children[3] = new BText(ctx['item'])
      b4.children[4] = new BText(\` \`)
      b4.children[5] = new BText(ctx['item_value'])
      b4.children[6] = new BText(\`] \`)
    }
    return b1;
  }
}"
`;

exports[`t-foreach iterate, position 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
  }
  
  return ctx => {
    const b1 = new Block1();
    const [v2, k2, l2] = getValues(Array(5));
    const b3 = b1.children[0] = new BCollection(l2);
    ctx = Object.create(ctx);
    for (let i1 = 0; i1 < l2; i1++) {
      ctx[\`elem\`] = v2[i1];
      ctx[\`elem_first\`] = i1 === 0;
      ctx[\`elem_last\`] = i1 === v2.length - 1;
      ctx[\`elem_index\`] = i1;
      ctx[\`elem_value\`] = k2[i1];
      const b4 = b3.children[i1] = new BMulti(6);
      b4.children[0] = new BText(\` -\`)
      if (ctx['elem_first']) {
        b4.children[1] = new BText(\` first\`)
      }
      if (ctx['elem_last']) {
        b4.children[2] = new BText(\` last\`)
      }
      b4.children[3] = new BText(\` (\`)
      b4.children[4] = new BText(ctx['elem_index'])
      b4.children[5] = new BText(\`) \`)
    }
    return b1;
  }
}"
`;

exports[`t-foreach multiple calls to t-raw 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/><div>Greeter</div><owl-anchor/></div>\`);
    children = new Array(2);
  }
  
  return ctx => {
    const b1 = new Block1();
    b1.children[0] = new BHtml(ctx[zero]);
    b1.children[1] = new BHtml(ctx[zero]);
    return b1;
  }
}"
`;

exports[`t-foreach multiple calls to t-raw 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
  }
  
  class Block2 extends BNode {
    static el = elem(\`<span>coucou</span>\`);
  }
  
  return (ctx, refs = {}) => {
    ctx = Object.create(ctx);
    ctx[scope] = 1;
    const b1 = new Block1();
    ctx = Object.create(ctx);
    const b2 = new Block2();
    ctx[zero] = b2;
    b1.children[0] = call(\`sub\`, ctx, refs);
    b1.refs = refs;
    return b1;
  }
}"
`;

exports[`t-foreach simple iteration (in a node) 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
  }
  
  return ctx => {
    const b1 = new Block1();
    const [v2, k2, l2] = getValues([3,2,1]);
    const b3 = b1.children[0] = new BCollection(l2);
    ctx = Object.create(ctx);
    for (let i1 = 0; i1 < l2; i1++) {
      ctx[\`item\`] = v2[i1];
      ctx[\`item_first\`] = i1 === 0;
      ctx[\`item_last\`] = i1 === v2.length - 1;
      ctx[\`item_index\`] = i1;
      ctx[\`item_value\`] = k2[i1];
      b3.children[i1] = new BText(ctx['item'])
    }
    return b1;
  }
}"
`;

exports[`t-foreach simple iteration 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  return ctx => {
    const [v1, k1, l1] = getValues([3,2,1]);
    const b2 = new BCollection(l1);
    ctx = Object.create(ctx);
    for (let i1 = 0; i1 < l1; i1++) {
      ctx[\`item\`] = v1[i1];
      ctx[\`item_first\`] = i1 === 0;
      ctx[\`item_last\`] = i1 === v1.length - 1;
      ctx[\`item_index\`] = i1;
      ctx[\`item_value\`] = k1[i1];
      b2.children[i1] = new BText(ctx['item'])
    }
    return b2;
  }
}"
`;

exports[`t-foreach t-call with body in t-foreach in t-foreach 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  return ctx => {
    const b1 = new BMulti(7);
    b1.children[0] = new BText(\` [\`)
    b1.children[1] = new BText(ctx['a'])
    b1.children[2] = new BText(\`] [\`)
    b1.children[3] = new BText(ctx['b'])
    b1.children[4] = new BText(\`] [\`)
    b1.children[5] = new BText(ctx['c'])
    b1.children[6] = new BText(\`] \`)
    return b1;
  }
}"
`;

exports[`t-foreach t-call with body in t-foreach in t-foreach 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/><span>[<owl-text/>][<owl-text/>][<owl-text/>]</span></div>\`);
    children = new Array(1);
    data = new Array(3);
    update() {
      const el1 = this.anchors[0].nextSibling.firstChild.nextSibling;
      const el2 = el1.nextSibling.nextSibling;
      el1.textContent = toString(this.data[0]);
      el2.textContent = toString(this.data[1]);
      el2.nextSibling.nextSibling.textContent = toString(this.data[2]);
    }
  }
  
  class Block2 extends BNode {
    static el = elem(\`<span><owl-text/></span>\`);
    data = new Array(1);
    update() {
      this.el.firstChild.textContent = toString(this.data[0]);
    }
  }
  
  return (ctx, refs = {}) => {
    ctx = Object.create(ctx);
    ctx[scope] = 1;
    const b1 = new Block1();
    const [v2, k2, l2] = getValues(ctx['numbers']);
    const b3 = b1.children[0] = new BCollection(l2);
    ctx = Object.create(ctx);
    for (let i1 = 0; i1 < l2; i1++) {
      ctx[\`a\`] = v2[i1];
      ctx[\`a_first\`] = i1 === 0;
      ctx[\`a_last\`] = i1 === v2.length - 1;
      ctx[\`a_index\`] = i1;
      ctx[\`a_value\`] = k2[i1];
      const b4 = b3.children[i1] = new BMulti(2);
      const [v5, k5, l5] = getValues(ctx['letters']);
      const b6 = b4.children[0] = new BCollection(l5);
      ctx = Object.create(ctx);
      for (let i2 = 0; i2 < l5; i2++) {
        ctx[\`b\`] = v5[i2];
        ctx[\`b_first\`] = i2 === 0;
        ctx[\`b_last\`] = i2 === v5.length - 1;
        ctx[\`b_index\`] = i2;
        ctx[\`b_value\`] = k5[i2];
        ctx = Object.create(ctx);
        ctx[\`c\`] = 'x'+'_'+ctx['a']+'_'+ctx['b'];
        b6.children[i2] = call(\`sub\`, ctx, refs);
        ctx = ctx.__proto__;
      }
      ctx = ctx.__proto__;
      const b7 = b4.children[1] = new Block2();
      b7.data[0] = ctx['c'];
    }
    ctx = ctx.__proto__;
    b1.data[0] = ctx['a'];
    b1.data[1] = ctx['b'];
    b1.data[2] = ctx['c'];
    b1.refs = refs;
    return b1;
  }
}"
`;

exports[`t-foreach t-call without body in t-foreach in t-foreach 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  return ctx => {
    ctx = Object.create(ctx);
    const b1 = new BMulti(7);
    ctx[\`c\`] = 'x'+'_'+ctx['a']+'_'+ctx['b'];
    b1.children[0] = new BText(\` [\`)
    b1.children[1] = new BText(ctx['a'])
    b1.children[2] = new BText(\`] [\`)
    b1.children[3] = new BText(ctx['b'])
    b1.children[4] = new BText(\`] [\`)
    b1.children[5] = new BText(ctx['c'])
    b1.children[6] = new BText(\`] \`)
    return b1;
  }
}"
`;

exports[`t-foreach t-call without body in t-foreach in t-foreach 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/><span>[<owl-text/>][<owl-text/>][<owl-text/>]</span></div>\`);
    children = new Array(1);
    data = new Array(3);
    update() {
      const el1 = this.anchors[0].nextSibling.firstChild.nextSibling;
      const el2 = el1.nextSibling.nextSibling;
      el1.textContent = toString(this.data[0]);
      el2.textContent = toString(this.data[1]);
      el2.nextSibling.nextSibling.textContent = toString(this.data[2]);
    }
  }
  
  class Block2 extends BNode {
    static el = elem(\`<span><owl-text/></span>\`);
    data = new Array(1);
    update() {
      this.el.firstChild.textContent = toString(this.data[0]);
    }
  }
  
  return (ctx, refs = {}) => {
    ctx = Object.create(ctx);
    ctx[scope] = 1;
    const b1 = new Block1();
    const [v2, k2, l2] = getValues(ctx['numbers']);
    const b3 = b1.children[0] = new BCollection(l2);
    ctx = Object.create(ctx);
    for (let i1 = 0; i1 < l2; i1++) {
      ctx[\`a\`] = v2[i1];
      ctx[\`a_first\`] = i1 === 0;
      ctx[\`a_last\`] = i1 === v2.length - 1;
      ctx[\`a_index\`] = i1;
      ctx[\`a_value\`] = k2[i1];
      const b4 = b3.children[i1] = new BMulti(2);
      const [v5, k5, l5] = getValues(ctx['letters']);
      const b6 = b4.children[0] = new BCollection(l5);
      ctx = Object.create(ctx);
      for (let i2 = 0; i2 < l5; i2++) {
        ctx[\`b\`] = v5[i2];
        ctx[\`b_first\`] = i2 === 0;
        ctx[\`b_last\`] = i2 === v5.length - 1;
        ctx[\`b_index\`] = i2;
        ctx[\`b_value\`] = k5[i2];
        b6.children[i2] = call(\`sub\`, ctx, refs);
      }
      ctx = ctx.__proto__;
      const b7 = b4.children[1] = new Block2();
      b7.data[0] = ctx['c'];
    }
    ctx = ctx.__proto__;
    b1.data[0] = ctx['a'];
    b1.data[1] = ctx['b'];
    b1.data[2] = ctx['c'];
    b1.refs = refs;
    return b1;
  }
}"
`;

exports[`t-foreach t-foreach in t-foreach 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
  }
  
  return ctx => {
    const b1 = new Block1();
    const [v2, k2, l2] = getValues(ctx['numbers']);
    const b3 = b1.children[0] = new BCollection(l2);
    ctx = Object.create(ctx);
    for (let i1 = 0; i1 < l2; i1++) {
      ctx[\`number\`] = v2[i1];
      ctx[\`number_first\`] = i1 === 0;
      ctx[\`number_last\`] = i1 === v2.length - 1;
      ctx[\`number_index\`] = i1;
      ctx[\`number_value\`] = k2[i1];
      const [v4, k4, l4] = getValues(ctx['letters']);
      const b5 = b3.children[i1] = new BCollection(l4);
      ctx = Object.create(ctx);
      for (let i2 = 0; i2 < l4; i2++) {
        ctx[\`letter\`] = v4[i2];
        ctx[\`letter_first\`] = i2 === 0;
        ctx[\`letter_last\`] = i2 === v4.length - 1;
        ctx[\`letter_index\`] = i2;
        ctx[\`letter_value\`] = k4[i2];
        const b6 = b5.children[i2] = new BMulti(4);
        b6.children[0] = new BText(\` [\`)
        b6.children[1] = new BText(ctx['number'])
        b6.children[2] = new BText(ctx['letter'])
        b6.children[3] = new BText(\`] \`)
      }
      ctx = ctx.__proto__;
    }
    return b1;
  }
}"
`;
